### PR TITLE
3961 migration address

### LINKED ISSUE #3961: Sprout-Sapling Migration Tool Step 2: Add the parameters (for the .conf file) for enabling the migration and setting the destination address

### PR DESCRIPTION
**Pull request by Eirik0**
Closes #3961

### CODE REVIEW

**Comment by Eirik0** [comment]
This includes the work from #3888 (currently 5 commits) and will be rebased after that is merged.

**Comment by Eirik0** [comment]
Force pushed with additional work from #3888.

**Comment by Eirik0** [comment]
Rebased on to master.

File: src/init.cpp [hunk 1]
```diff
@@ -1599,6 +1601,14 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
         LogPrintf("Wallet disabled!\n");
     } else {
+        // Check Sapling migration address if set and is a valid Sapling address
+        if (mapArgs.count("-migrationdestaddress")) {
+            std::string migrationDestAddress = mapArgs["-migrationdestaddress"];
+            libzcash::PaymentAddress address = DecodePaymentAddress(migrationDestAddress);
+            if (boost::get<libzcash::SaplingPaymentAddress>(&address) == nullptr) {
+                return InitError(_("-migrationdestaddress must be a valid Sapling address."));
+            }
+        }
```
Comment (str4d) [inline src/init.cpp]: Move this into the second `ENABLE_WALLET` section of Step 3.

File: src/init.cpp [hunk 2]
```diff
@@ -1671,6 +1681,9 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
             }
         }
 
+        // Set sapling migration status
+        pwalletMain->fSaplingMigrationEnabled = GetBoolArg("-migration", false);
```
Comment (str4d) [inline src/init.cpp]: Add this option to the help text.

**Comment by Eirik0** [comment]
@str4d force pushed with your suggested changes.

File: src/init.cpp [hunk 3]
```diff
@@ -405,6 +406,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageGroup(_("Wallet options:"));
     strUsage += HelpMessageOpt("-disablewallet", _("Do not load the wallet and disable wallet RPC calls"));
     strUsage += HelpMessageOpt("-keypool=<n>", strprintf(_("Set key pool size to <n> (default: %u)"), 100));
+    strUsage += HelpMessageOpt("--migration=<boolean>", _("Set to true to enable the Sprout to Sapling migration."));
```
Comment (str4d) [inline src/init.cpp]:
```suggestion
    strUsage += HelpMessageOpt("-migration=<boolean>", _("Set to true to enable the Sprout to Sapling migration."));
```

**Comment by Eirik0** [comment]
I added tests for this as part of https://github.com/zcash/zcash/pull/3973. It's not ideal to have the disconnect, but it just seemed easier to test once we had the call to get the migration status.

**Comment by Eirik0** [comment]
@zkbot r+

**Pull request merged**